<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: MSKCC Bioinformatics Core RNASeq</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> MSKCC Bioinformatics Core RNASeq
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;bicrnaseq&rsquo; version 0.0.1</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="bic.add.rownames2dat.html">bic.add.rownames2dat</a></td>
<td>Add a column to matrix containing its row names</td></tr>
<tr><td style="width: 25%;"><a href="bic.average.by.name.html">bic.average.by.name</a></td>
<td>Average all values in a matrix that have the same 'name'</td></tr>
<tr><td style="width: 25%;"><a href="bic.check.line.chart.data.html">bic.check.line.chart.data</a></td>
<td>Check formatting of input to bic.plot.rseqc.line.chart()</td></tr>
<tr><td style="width: 25%;"><a href="bic.check.picard.data.html">bic.check.picard.data</a></td>
<td>Validate data from PICARD CollectRnaSeqMetrics or AlignmentSummaryMetrics</td></tr>
<tr><td style="width: 25%;"><a href="bic.check.read.distribution.data.html">bic.check.read.distribution.data</a></td>
<td>Check formatting of input to bic.plot.read.distribution()</td></tr>
<tr><td style="width: 25%;"><a href="bic.deseq.heatmap.html">bic.deseq.heatmap</a></td>
<td>Draw a heatmap of counts table</td></tr>
<tr><td style="width: 25%;"><a href="bic.deseq.normalize.htseq.counts.html">bic.deseq.normalize.htseq.counts</a></td>
<td>Use DESeq to normalize raw HTSeq counts</td></tr>
<tr><td style="width: 25%;"><a href="bic.deseq.plot.pca.html">bic.deseq.plot.pca</a></td>
<td>Plot PCA</td></tr>
<tr><td style="width: 25%;"><a href="bic.file2matrix.html">bic.file2matrix</a></td>
<td>Read data file and return as a matrix</td></tr>
<tr><td style="width: 25%;"><a href="bic.format.htseq.counts.html">bic.format.htseq.counts</a></td>
<td>Read in raw HTSeq count file and format for analysis</td></tr>
<tr><td style="width: 25%;"><a href="bic.get.deseq.cds.html">bic.get.deseq.cds</a></td>
<td>Normalize raw count matrix using DESeq scaling method</td></tr>
<tr><td style="width: 25%;"><a href="bic.get.gene.sets.html">bic.get.gene.sets</a></td>
<td>Get the gene sets that meet criteria set by cut offs</td></tr>
<tr><td style="width: 25%;"><a href="bic.get.package.info.html">bic.get.package.info</a></td>
<td>Get info for one or multiple packages installed in current R version</td></tr>
<tr><td style="width: 25%;"><a href="bic.grid.arrange.shared.legend.html">bic.grid.arrange.shared.legend</a></td>
<td>Arrange multiple plots on one page with one shared legend</td></tr>
<tr><td style="width: 25%;"><a href="bic.hclust.samples.html">bic.hclust.samples</a></td>
<td>Plot heirachical clustering of all samples</td></tr>
<tr><td style="width: 25%;"><a href="bic.join.strings.html">bic.join.strings</a></td>
<td>Convert a character vector to a delimited string</td></tr>
<tr><td style="width: 25%;"><a href="bic.make.colnames.html">bic.make.colnames</a></td>
<td>Remove a row from matrix to use the values as column names</td></tr>
<tr><td style="width: 25%;"><a href="bic.make.rownames.html">bic.make.rownames</a></td>
<td>Remove a column from a matrix and use its values as the matrix's rownames.</td></tr>
<tr><td style="width: 25%;"><a href="bic.matrix2numeric.html">bic.matrix2numeric</a></td>
<td>Convert all values in matrix to numeric</td></tr>
<tr><td style="width: 25%;"><a href="bic.mds.clust.samples.html">bic.mds.clust.samples</a></td>
<td>Plot MDS clustering of all samples</td></tr>
<tr><td style="width: 25%;"><a href="bic.non.empty.file.found.html">bic.non.empty.file.found</a></td>
<td>Search for a non-empty file by pattern</td></tr>
<tr><td style="width: 25%;"><a href="bic.pdf.hclust.html">bic.pdf.hclust</a></td>
<td>Write PDF file containing heierarchical clustering tree</td></tr>
<tr><td style="width: 25%;"><a href="bic.plot.5prime3prime.bias.html">bic.plot.5prime3prime.bias</a></td>
<td>Plot coverage uniformity (5' and 3' bias)</td></tr>
<tr><td style="width: 25%;"><a href="bic.plot.alignment.distribution.html">bic.plot.alignment.distribution</a></td>
<td>Plot distibution of alignment across different genomic locations</td></tr>
<tr><td style="width: 25%;"><a href="bic.plot.alignment.summary.html">bic.plot.alignment.summary</a></td>
<td>Plot numbers of mapped and unmapped reads for each sample</td></tr>
<tr><td style="width: 25%;"><a href="bic.plot.coverage.html">bic.plot.coverage</a></td>
<td>Plot normalized coverage</td></tr>
<tr><td style="width: 25%;"><a href="bic.plot.dispersion.estimates.html">bic.plot.dispersion.estimates</a></td>
<td>Plot dispersion estimates</td></tr>
<tr><td style="width: 25%;"><a href="bic.plot.ma.html">bic.plot.ma</a></td>
<td>Plot MA</td></tr>
<tr><td style="width: 25%;"><a href="bic.plot.read.distribution.html">bic.plot.read.distribution</a></td>
<td>Plot RSeQC read distribution stats</td></tr>
<tr><td style="width: 25%;"><a href="bic.plot.rseqc.line.chart.html">bic.plot.rseqc.line.chart</a></td>
<td>Plot RSeQC sequential metrics in a line graph</td></tr>
<tr><td style="width: 25%;"><a href="bic.process.gsa.res.html">bic.process.gsa.res</a></td>
<td>Format and filter results from piano::runGSA()</td></tr>
<tr><td style="width: 25%;"><a href="bic.pval.histogram.html">bic.pval.histogram</a></td>
<td>Histogram of p-values</td></tr>
<tr><td style="width: 25%;"><a href="bic.quantile.normalize.htseq.counts.html">bic.quantile.normalize.htseq.counts</a></td>
<td>Use quantile normalization to normalize raw HTSeq counts</td></tr>
<tr><td style="width: 25%;"><a href="bic.remove.sub.html">bic.remove.sub</a></td>
<td>Remove a substring from a string</td></tr>
<tr><td style="width: 25%;"><a href="bic.run.deseq.comparison.html">bic.run.deseq.comparison</a></td>
<td>Run a differential expression comparison between two conditions.</td></tr>
<tr><td style="width: 25%;"><a href="bic.run.gsa.html">bic.run.gsa</a></td>
<td>Run gene set analysis using the package 'piano'</td></tr>
<tr><td style="width: 25%;"><a href="bic.sample.to.sample.distances.html">bic.sample.to.sample.distances</a></td>
<td>Draw a heatmap showing sample-to-sample distances using variance stabilized data</td></tr>
<tr><td style="width: 25%;"><a href="bic.scale.factor.quantile.html">bic.scale.factor.quantile</a></td>
<td>Non-deseq way to normalize counts(?)</td></tr>
<tr><td style="width: 25%;"><a href="bic.standard.heatmap.html">bic.standard.heatmap</a></td>
<td>Create PDF of generic red/green/black heatmap showing relative expression values</td></tr>
<tr><td style="width: 25%;"><a href="bic.write.dat.html">bic.write.dat</a></td>
<td>Write file containing tab delimited data</td></tr>
<tr><td style="width: 25%;"><a href="bic.write.deseq.results.html">bic.write.deseq.results</a></td>
<td>Write results for ALL genes and for DE genes</td></tr>
<tr><td style="width: 25%;"><a href="bic.write.normalized.counts.file.html">bic.write.normalized.counts.file</a></td>
<td>Write normalized counts matrix to file</td></tr>
</table>
</body></html>
